include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td

```diff
@@ -283,14 +286,17 @@ def AndOfPad : Pat <
   [(KnownWidth $x), (UIntType $x), (EqualTypes $x, $pad)]>;
 
 /// cvt is adding a 0, which will pass through the and
-/// and(cvt(x:U), const) -> cat(0, and(x,trunc(const))
+/// Only trigger when cvt is full length
+/// and(cvt(x:U), const) -> cat(0, and(x,sextortrunc(const))
 def AndCvtU : Pat <
-  (AndPrimOp:$old (CvtPrimOp $x), (ConstantOp:$ocst $cst)),
+  (AndPrimOp:$old (CvtPrimOp:$cvt $x), (ConstantOp:$ocst $cst)),
   (MoveNameHint $old, (CatPrimOp 
     (NativeCodeCall<"$_builder.create<ConstantOp>($0.getLoc(), APSInt(APInt()))"> $ocst),
-    (AndPrimOp $x, (NativeCodeCall<"$_builder.create<ConstantOp>($0.getLoc(), cast<IntType>($0.getType()), $1.getValue().trunc($1.getValue().getBitWidth()-1).zext(*cast<IntType>($0.getType()).getWidth()))"> $x, $cst))
+    (AndPrimOp $x, (NativeCodeCall<"$_builder.create<ConstantOp>($0.getLoc(), "
+                                      "cast<IntType>($0.getType()), "
+                                      "$1.getValue().sextOrTrunc(cast<IntType>($0.getType()).getWidthOrSentinel()))"> $x, $cst))
```

Comment:
:thinking: could this pattern work for non-constants (of known width) equally well, and let a different folder create the `sextortrunc` thing for us?

Reply:
(This can be done separately, partly just curious, it's of course more important to fix the issue, thanks!)

Reply:
I was playing with a conversion via asSInt and then a width-narrowing pattern.

Reply:
We have more robust width-narrowing in comb, and I'm inclined to do more of this stuff in comb where it's less error prone.